synapse/storage/databases/main/room.py

```diff
@@ -1148,22 +1148,24 @@ def _quarantine_media_txn(
         """
 
         # Update all the tables to set the quarantined_by flag
+        # We also pick up any media with a matching hash.
         sql = """
             UPDATE local_media_repository
             SET quarantined_by = ?
-            WHERE media_id = ?
+            WHERE (media_id = ?
+            OR sha256 IN (SELECT DISTINCT sha256 FROM local_media_repository WHERE media_id = ?))
```

Comment:
```suggestion
            OR sha256 IN (SELECT DISTINCT sha256 FROM local_media_repository WHERE media_id = ? AND sha256 IS NOT NULL))
```

I am not convinced this is necessary (since `NULL IN (..., NULL, ...)` should evaluate to `NULL`, which is non-truthy in a `WHERE` condition), but I prefer it